torch/utils/_content_store.py

```diff
@@ -0,0 +1,156 @@
+# assume that you can keep all of the tensors you want to add to the store
+# in memory at once, because you probably can't!  Nor do we assume that
+# you know a priori whether or not two storages can be deduplicated or not.
+#
+# Note: only storages are content-addressed; tensors are name addressed
+#
+# Note: our padding strategy means that [1, 0] and [1] int16 tensors would
+# map to the same (padded) storage.  We think this will be immaterial for most
+# users.
+
+import os.path
+import struct
+from typing import Optional
+
+import torch
+import torch._prims as prims
+import torch._utils
+import torch.nn.functional as F
+
+from torch.multiprocessing.reductions import StorageWeakRef
+
+
+def hash_storage(storage):
+    # Use of torch.compile is mandatory for (1) good memory usage
+    # and (2) xor_sum implementation
+    @torch.compile(dynamic=True)
+    def kernel(x):
+        # The randint calls are carefully written to hit things we
+        # have lowerings for in inductor.  Lack of unsigned 32-bit integer
+        # is a pain.
+        a = torch.randint(
+            -(2**31), 2**31, x.shape, device=x.device, dtype=torch.int32
+        ).abs()
+        a = ((a % (2**31 - 1)) + 1).long()
+        b = (
+            torch.randint(
+                -(2**31), 2**31, x.shape, device=x.device, dtype=torch.int32
+            )
+            .abs()
+            .long()
+        )
+        # This is a standard shift-multiply universal hash family
+        # plus xor sum hash, using Philox to generate random numbers.
+        # Our Philox RNG is not deterministic across devices so
+        # don't use this for stable hashing.
+        #
+        # This assumes fixed length so you're also obligated to bucket
+        # by the length of tensor as well
+        return prims.xor_sum((a * x + b).int(), [0])
+
+    with torch.random.fork_rng(
+        [storage.device] if storage.device.type != "cpu" else []
+    ):
+        torch.manual_seed(0)  # this can be anything, just needs to be fixed
+        x = torch.empty(0, dtype=torch.uint8, device=storage.device).set_(storage)
+        # The dtype-casting view cannot be compiled, and so the
+        # padding/reshaping also needs to be done externally even
+        # though it could be profitably fused
+        pad = x.numel() % 4
+        if pad > 0:
+            x = F.pad(x, (0, pad), "constant", 0)
+        x = x.view(torch.int32)
+        torch._dynamo.mark_dynamic(x, 0)
+        # We run the 32-bit hash five times with differing parameters to
+        # reduce chance of collision
+        ITER = 5
+        cs = [kernel(x).item() for _ in range(ITER)]
+        return struct.pack(">" + "i" * ITER, *cs).hex()
+
+
+class ContentStoreWriter:
+    # Structure:
+    #   storages/
+    #     00/
+    #       0000..00
+    #   tensors/
+    #     name
+    def __init__(self, loc):
+        self.loc = loc
+        self.seen_storage_hashes = set()
+
+    def write_storage(self, storage):
```

Comment:
prefix `_`? Seems like write_tensor is the API, not write_storage?

Reply:
I actually use write_storage directly in the next PR, they're both valid APIs.